File: oss/clawpipe/dotnet-sdk/ClawPipe/Packer.cs
using System.Text.RegularExpressions;

namespace ClawPipe;

/// <summary>
/// Context Packer — compress prompt text to reduce token count.
/// Strategies: whitespace compression, deduplication, boilerplate stripping.
/// </summary>
public sealed partial class Packer
{
    private readonly PackerOptions _options;

    public Packer(PackerOptions? options = null) => _options = options ?? new();

    /// <summary>Pack a prompt, returning compressed text and savings percentage.</summary>
    public PackResult Pack(string text)
    {
        var original = text;
        var originalTokens = EstimateTokens(original);

        var packed = original;
        if (_options.CompressWhitespace) packed = CompressWhitespace(packed);
        if (_options.Deduplicate) packed = Deduplicate(packed);
        if (_options.StripBoilerplate) packed = StripBoilerplate(packed);
        packed = TruncateToLimit(packed);

        var packedTokens = EstimateTokens(packed);
        var savingsPct = originalTokens > 0
            ? Math.Max(0.0, (1.0 - (double)packedTokens / originalTokens) * 100)
            : 0.0;

        return new PackResult(packed, Math.Round(savingsPct, 1));
    }

    /// <summary>Rough token estimate: ~4 characters per token.</summary>
    public static int EstimateTokens(string text) => (int)Math.Ceiling(text.Length / 4.0);

    // ── private helpers ────────────────────────────────────────────────────

    private static string CompressWhitespace(string text)
    {
        var lines = text.Split('\n').Select(l => l.TrimEnd());
        var joined = string.Join('\n', lines);
        return MultipleNewlines().Replace(joined, "\n\n").Trim();
    }

    private static string Deduplicate(string text)
    {
        var blocks = text.Split("\n\n");
        var seen = new HashSet<string>();
        var unique = new List<string>();
        foreach (var block in blocks)
        {
            var normalized = block.Trim().ToLowerInvariant();
            if (string.IsNullOrEmpty(normalized)) continue;
            if (normalized.Length > 50 && seen.Contains(normalized)) continue;
            seen.Add(normalized);
            unique.Add(block);
        }
        return string.Join("\n\n", unique);
    }

    [GeneratedRegex(@"^/\*\*?\s*\n(\s*\*\s*@(param|returns|throws|example).*\n)*\s*\*/", RegexOptions.Multiline)]
    private static partial Regex JsDocPattern();

    [GeneratedRegex(@"^//\s*eslint-disable.*$", RegexOptions.Multiline)]
    private static partial Regex EslintDisablePattern();

    [GeneratedRegex(@"^//\s*@ts-(ignore|expect-error|nocheck).*$", RegexOptions.Multiline)]
    private static partial Regex TsIgnorePattern();

    [GeneratedRegex(@"^'use strict';?\s*$", RegexOptions.Multiline)]
    private static partial Regex UseStrictPattern();

    [GeneratedRegex(@"^\n{3,}", RegexOptions.Multiline)]
    private static partial Regex MultipleNewlines();

    private static string StripBoilerplate(string text)
    {
        var result = JsDocPattern().Replace(text, "");
        result = EslintDisablePattern().Replace(result, "");
        result = TsIgnorePattern().Replace(result, "");
        result = UseStrictPattern().Replace(result, "");
        return MultipleNewlines().Replace(result, "\n\n").Trim();
    }

    private string TruncateToLimit(string text)
    {
        var maxChars = _options.MaxTokens * 4;
        if (text.Length <= maxChars) return text;
        var truncated = text[..maxChars];
        var lastNl = truncated.LastIndexOf('\n');
        var cut = lastNl > maxChars * 0.8 ? lastNl : maxChars;
        return truncated[..cut] + "\n\n[Truncated -- context exceeded budget]";
    }
}

/// <summary>Options for the Packer.</summary>
public sealed record PackerOptions(
    int MaxTokens = 100_000,
    bool Deduplicate = true,
    bool StripBoilerplate = true,
    bool CompressWhitespace = true
);

File: oss/clawpipe/elixir-sdk/lib/clawpipe/packer.ex
defmodule ClawPipe.Packer do
  @moduledoc """
  Context Packer — compress context to reduce token count.

  Strategies applied in order:
  1. Whitespace compression
  2. Paragraph deduplication
  3. Boilerplate stripping (eslint-disable, @ts-*, 'use strict', JSDoc tags)
  4. Budget truncation (default 100 000 tokens)
  """

  @boilerplate_patterns [
    # JSDoc blocks with only annotation lines
    ~r/\/\*\*?\s*\n(\s*\*\s*@(?:param|returns|throws|example).*\n)*\s*\*\//m,
    # eslint-disable comments
    ~r/^\/\/\s*eslint-disable.*$/m,
    # TypeScript suppression comments
    ~r/^\/\/\s*@ts-(?:ignore|expect-error|nocheck).*$/m,
    # 'use strict' directives
    ~r/^'use strict';?\s*$/m,
    # istanbul ignore annotations
    ~r/^\/\*\s*istanbul ignore (?:next|else)\s*\*\/$/m
  ]

  @default_max_tokens 100_000

  @doc """
  Pack a prompt, returning a map with `:text` and `:savings_pct`.

  Options:
    - `:max_tokens`        — token budget (default #{@default_max_tokens})
    - `:system`            — optional system message prepended before packing
    - `:compress_whitespace` — boolean (default true)
    - `:deduplication`     — boolean (default true)
    - `:strip_boilerplate` — boolean (default true)
  """
  @spec pack(String.t(), keyword()) :: %{text: String.t(), savings_pct: float()}
  def pack(text, opts \\ []) when is_binary(text) do
    max_tokens        = Keyword.get(opts, :max_tokens, @default_max_tokens)
    system            = Keyword.get(opts, :system)
    do_whitespace     = Keyword.get(opts, :compress_whitespace, true)
    do_dedup          = Keyword.get(opts, :deduplication, true)
    do_strip          = Keyword.get(opts, :strip_boilerplate, true)

    original = if system, do: "#{system}\n\n#{text}", else: text
    original_tokens = estimate_tokens(original)

    packed =
      original
      |> maybe_compress_whitespace(do_whitespace)
      |> maybe_deduplicate(do_dedup)
      |> maybe_strip_boilerplate(do_strip)
      |> truncate(max_tokens)

    packed_tokens = estimate_tokens(packed)

    savings_pct =
      if original_tokens > 0 do
        Float.round((1 - packed_tokens / original_tokens) * 100, 1)
      else
        0.0
      end

    %{text: packed, savings_pct: max(0.0, savings_pct)}
  end

  @doc "Rough token estimate: ~4 chars per token."
  @spec estimate_tokens(String.t()) :: non_neg_integer()
  def estimate_tokens(text) when is_binary(text) do
    ceil(String.length(text) / 4)
  end

  # --- Internals ---

  defp maybe_compress_whitespace(text, true), do: compress_whitespace(text)
  defp maybe_compress_whitespace(text, _), do: text

  defp maybe_deduplicate(text, true), do: deduplicate(text)
  defp maybe_deduplicate(text, _), do: text

  defp maybe_strip_boilerplate(text, true), do: strip_boilerplate(text)
  defp maybe_strip_boilerplate(text, _), do: text

  defp compress_whitespace(text) do
    text
    |> String.split("\n")
    |> Enum.map(&String.trim_trailing/1)
    |> Enum.join("\n")
    |> then(&Regex.replace(~r/\n{3,}/, &1, "\n\n"))
    |> String.trim()
  end

  defp deduplicate(text) do
    text
    |> String.split("\n\n")
    |> Enum.reduce({[], MapSet.new()}, fn block, {acc, seen} ->
      normalized = block |> String.trim() |> String.downcase()

      if normalized == "" do
        {acc, seen}
      else
        if String.length(normalized) > 50 and MapSet.member?(seen, normalized) do
          {acc, seen}
        else
          {[block | acc], MapSet.put(seen, normalized)}
        end
      end
    end)
    |> then(fn {acc, _} -> acc |> Enum.reverse() |> Enum.join("\n\n") end)
  end

  defp strip_boilerplate(text) do
    result =
      Enum.reduce(@boilerplate_patterns, text, fn pat, acc ->
        Regex.replace(pat, acc, "")
      end)

    Regex.replace(~r/\n{3,}/, result, "\n\n") |> String.trim()
  end

  defp truncate(text, max_tokens) do
    max_chars = max_tokens * 4

    if String.length(text) <= max_chars do
      text
    else
      truncated = String.slice(text, 0, max_chars)
      last_nl   = find_last_newline(truncated, max_chars)
      cut       = if last_nl > max_chars * 0.8, do: last_nl, else: max_chars

      String.slice(text, 0, cut) <> "\n\n[Truncated — context exceeded budget]"
    end
  end

  defp find_last_newline(text, from) do
    case :binary.match(text, "\n", scope: {0, from}) do
      :nomatch -> 0
      {pos, _}  -> pos
    end
  end
end
